module: resolve and instantiate loader pipeline hooks

guybedford opened the pull request:
This PR provides an implementation of a hooking mechanism worked on with some early feedback from @bmeck.

#### Usage

The es module loading pipeline is hooked via a `--loader` flag to Node, which specifies a file containing a loader definition.

For example usage:

* A loader that loads all files (including `.js`) as es modules is provided at https://github.com/guybedford/node/blob/resolve-hook-rebased/test/fixtures/es-module-loaders/js-loader.mjs.
* A loader that provides named exports for NodeJS builtin modules is provided at https://github.com/guybedford/node/blob/resolve-hook-rebased/test/fixtures/es-module-loaders/builtin-named-exports-loader.mjs.

#### Motivation

This work is built to follow principles of extensibility - allowing users (framework and platform developers for this feature) to customize their own hooks for their use case needs. The core ecosystem conventions for loading modules in NodeJS will remain the ones that NodeJS sets itself by the ecosystem popularity due to interop requirements.

The goal here though is that through having the ability for users to experiment with loader features will allow the most useful and popular features to potentially be merged back into NodeJS core to extend the ecosystem conventions. So roughly the extensible manifesto thinking.

I believe enabling these use cases is important to avoid possible ecosystem fragmentation through other means.

The examples above show common use cases, and these hooks are also sufficient to implement the package.json `"module"` property resolution concept using a loader (https://github.com/nodejs/node-eps/pull/60).

#### Architecture

The hook architecture here is inspired by learnings from the WhatWG loader work, while aiming to cut down the scope of hooking to a bare minimum API surface area, as can be seen from the examples.

The loader itself is given control of module resolution, of the form:

```ts
resolve (specifier: string, parentModuleUrl: string) => { url: string, format: string }
```

Where the `"format"` returned is `"esm"` for the ES module loader, `"cjs"` for the traditional cjs loader, `"builtin"` to refer to a NodeJS core module, `"json"` for json and `"addon"` for a Node binary addon. In future this could be extended for new module formats (ie wasm, binary ast).

This resolver follows exactly what has always been the resolve hook in the module loader specifications, except instead of just returning a `string` URL, it returns the URL along with the format. Other information could also possibly be returned on this object in future including for example integrity metadata.

When the format returned is `"dynamic"`, then we delegate loading to a `dynamicInstantiate` hook, hooked in the same way as `resolve`:

```ts
async dynamicInstantiate (url: string) => ({
  exports: string[],
  execute: (exports) => {
    // get and set functions provided for pre-allocated export names
    exports.x.set('asdf');
  }
})
```

This API is designed to enable the above example use cases, while ensuring that loader internals are kept private, to ensure that what is hooked is only what is intended to be hooked. For example, none of `ModuleWrap` is exposed.

Feedback very welcome!

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/CONTRIBUTING.md#commit-message-guidelines)

##### Affected core subsystem(s)
<!-- Provide affected core subsystem(s) (like doc, cluster, crypto, etc). -->
esmodules

bcoe commented:
@guybedford what might it look like to extend on the hooks available, such that a modified source-file could be stored in the cache?

I'm interested in how we'll continue providing test-coverage for `.mjs` files (which I believe circumvent the current `require.extensions` API -- this seems like it could be a step in the direction of what I'd need to get Istnabul working with `.mjs`?

guybedford commented:
@bcoe one option might be to add a separate 'fetch' hook to such a loader object alongside the 'resolve' here which could then add instrumentation. Happy to prototype it if there's interest.

bcoe commented:
@guybedford I'm _very_ interested. The built-in coverage in V8 doesn't seem to quite be over the finish line yet, so this would be a great transitional step for continuing to support test-coverage with Istanbul.

bcoe commented:
@guybedford @bmeck any chance we could also allow you to attach a loader in a way similar to `module.extensions`. Perhaps `module.loader`? This would make the implementation way easier for tools like Istanbul. Then I could simply check whether `module.loader` exists, and instrument code accordingly.

guybedford commented:
@bcoe something like `module.loaderHooks = { fetch, resolve }` or `module.registerHook('resolve', resolve)` is possible, but there is the async load timing issue here in that the boot of the application will not run through the hooks in that case. The boot for such a scenario then might look like:

```js
import module from 'module';
import { customResolve, customFetch } from 'custom-loader';

module.registerHook('resolve', customResolve);
module.registerHook('fetch', customFetch);

// now kick of the dynamic import hooking the main application tree load
import('./entry-point.mjs');
```

This was actually roughly what I had for the original API here, but after discussion with @bmeck we adopted a flag rather here to avoid some of the above issues. I'm still open to either approach.

jasnell commented on lib/internal/errors.js:
```diff
@@ -287,6 +287,8 @@ E('ERR_TRANSFORM_WITH_LENGTH_0',
 E('ERR_UNESCAPED_CHARACTERS',
   (name) => `${name} contains unescaped characters`);
 E('ERR_UNKNOWN_ENCODING', 'Unknown encoding: %s');
+E('ERR_UNKNOWN_FILE_EXTENSION', 'Uknown file extension: %s');
```
typo: s/Uknown/Unknown

Fishrock123 replied:
Maybe this could print out the file too?

guybedford replied:
Sure, instead of just outputting the extension we can output the full path.

jasnell commented on lib/internal/errors.js:
```diff
@@ -287,6 +287,8 @@ E('ERR_TRANSFORM_WITH_LENGTH_0',
   (name) => `${name} contains unescaped characters`);
 E('ERR_UNKNOWN_ENCODING', 'Unknown encoding: %s');
+E('ERR_UNKNOWN_FILE_EXTENSION', 'Uknown file extension: %s');
+E('ERR_UNKNOWN_LOADER', 'Uknown module loader: %s');
```
here too :-)

Fishrock123 replied:
The message here seems a bit vague - what does this mean?

jasnell commented on src/node.cc:
```diff
@@ -3950,6 +3955,8 @@ static void CheckIfAllowedInEnv(const char* exe, bool is_env,
     "--napi-modules",
     "--expose-http2",
+    "--experimental-modules",
+    "--loader",
```
Perhaps `--module-loader` would be more descriptive?

Fishrock123 replied:
@jasnell what might it be confused by?

jasnell commented on src/node_config.cc:
```diff
@@ -65,8 +65,18 @@ static void InitConfig(Local<Object> target,
-  if (config_experimental_modules)
+  if (config_experimental_modules) {
     READONLY_BOOLEAN_PROPERTY("experimentalModules");
+    if (!config_userland_loader.empty()) {
```
thank you for using `node_config` to expose this :-)

guybedford replied:
That is all @bmeck!

jasnell commented:
Awesome to see this @guybedford ... left a few initial review comments but will need to play with this a bit more before signing off. Thank you for working on this

bmeck commented:
I'd like to leave things as declarative instead of API based if possible for now. `require.extensions` and people monkey patching `module.*` have cause a lot of lock-in that we cannot undo in CJS in part because of how hard it is to tell when something is being used, and in part because of the power of making them JS object properties.

As for a fetch transform, I have several ideas on how to accomplish similar ideas without introducing that transform:

1. make a file on disk w/ mutated source and resolve to that.
2. we introduce blob backed `URL.createObjectURL` and use that during resolution.

These can be worked through in the browser if done carefully; I am not convinced we should introduce fetch hooks in Node without creating a story for network backed modules and security of them. I also would like some involvement from browser people who are creating multiple different execution workers like ServiceWorker in order to do these tasks.

For code coverage I know V8 now has that instrumented in the codegen, so perhaps using that and adding missing features would be easier over time than trying to have hooks. Especially since we don't have a composition mechanism for these hooks currently.

This PR looks good even though I think we have more than enough hook area already and should not expand the scope of the PR. Overall, I think minimal hooks is the best approach in the beginning as I still think these hooks will change as we use them / won't look exactly the same by the time they are released.

feature | opinion | notes
---- | ---- | ----
Resolve hook | +1 | need composition eventually
Evaluation hook | +1 | would like more research over time into properly managing live bindings for core (e.g. https://github.com/bmeck/node/tree/named-export-core using a Proxy)
Fetch hook | -1 | seems like alternates
JS based API | -1 | has been a source of pain in past, no clear use case need for it vs declarative

domenic commented:
I'm not sure when this is targeted to land, but this seems very early. I would hope this doesn't end up in Node core at least until the ecosystem has had some time to settle down and explore alternatives and prove that this particular API shape is the best, and that the use cases it addresses are actually important enough to add such a powerful, fragmentation-encouraging primitive which cannot be taken back.

General API comment: please never return or accept URL objects; see https://url.spec.whatwg.org/#url-apis-elsewhere .

On the actual capabilities here, I think the resolve hook is straightforward enough (although may end up mismatching browsers if they need something more declarative, as is currently the thinking). The loader object is much scarier and is nowhere near what we would currently contemplate exposing in browsers (as essentially we are trying to stick with Source Text Module Records, so any hook I can imagine would be just a fetch hook, i.e. service workers).

Fishrock123 commented on lib/internal/loader/ModuleRequest.js:
```diff
@@ -0,0 +1,87 @@
+
+// Strategy for loading a standard JavaScript module
+loaders.set('esm', async (url) => {
+  const source = `${await asyncReadFile(url)}`;
```
What happens if the readFile throws?

bmeck replied:
1. The loader fails to produce a ModuleWrap
2. The module job it is placed in produces an error (cached as per spec): https://github.com/nodejs/node/pull/15445/files/db6eb0069cfc94d00b647525a47ddc73dba9ccd8#diff-e6f2798544dae1f6ec863621e745489aR93 
3. All imports (current and in future) of that absolute url fail due to the job failing:  https://github.com/nodejs/node/pull/15445/files/db6eb0069cfc94d00b647525a47ddc73dba9ccd8#diff-e6f2798544dae1f6ec863621e745489aR101

Fishrock123 commented on lib/internal/loader/ModuleRequest.js:
```diff
@@ -0,0 +1,87 @@
+      return { url, loader: 'esm' };
+    case '.json':
+      return { url, loader: 'cjs' };
+    case '.node':
```
Aren't `.node` files native addon C++?

bmeck replied:
they use the CJS interop

guybedford replied:
It might be nice to add a `binary` format now - suggestions welcome for how to load a file as _just_ a Node binary, throwing otherwise.

Fishrock123 commented on lib/internal/loader/ModuleRequest.js:
```diff
@@ -0,0 +1,87 @@
+    case '.js':
+      return { url, loader: 'cjs' };
+    default:
+      throw new errors.Error('ERR_UNKNOWN_FILE_EXTENSION', ext);
```
Shouldn't this default to `'cjs'`? (Though I suppose this is not required for importing CJS via ESM...)

bmeck replied:
This is for future proofing, just like https://github.com/nodejs/node/pull/15365 was talking about.

guybedford commented:
Thanks so much for all feedback so far. I've updated the description above to match the latest updates:

* I've renamed `loader` to `format` in the resolver hook, which is now fully consistent with being interpreted as the _module format_ of the module being loaded, and no longer accepts a loader function. Hopefully this simplifies the resolve hook and makes it easier to see it as something that might be compatible with browsers too.
* To support the previous loader option, when the `"dynamic"` format is returned, module loading is delegated to a separate `dynamicInstantiate` hook which takes the place of the functional loader option.
* The hooks no longer accept or return URL objects, but work with string URLs only.

benjamn commented on lib/module.js:
```diff
@@ -490,6 +490,10 @@ function tryModuleLoad(module, filename) {
 
+Module.import = function(request, parent) {
+  return ESMLoader.import(request, parent);
+};
```
Why not make this `Module.prototype.import` so you can get rid of the `parent` parameter (and use `this` instead)?

bmeck replied:
there eventually could be multiple loaders / I would like to see this fn removed when `import` lands

jdalton replied:
`Module.import` doesn't appear to be used anywhere else in this PR, is it dead code?

bmeck replied:
used as stop gap til `import()` in 6.2 lands

jdalton replied:
> used as stop gap til import() in V8@6.2 lands

Is there a rough idea of when that'll be?
Would `Module.import` be intended as public consumable API?

bmeck replied:
Sometime in Node@9
no, i'd remove it once 6.2 lands.

bmeck replied:
We should guard this by checking for experimentalModules

iarna commented:
Ok, so @bcoe asked me to chime in:

My primary use case for a custom loader is _mocks_. I currently maintain [require-inject](https://www.npmjs.com/package/require-inject) which fiddles with the cache to allow tests to replace modules with their own code. How it does this is not something I'd want to see replicated as it's very much a hack, but the functionality is something I don't want to lose with ESM.

It looks like a loader API + `Module.import()` or a v8 `import()` would meet my needs.

An interpreter flag might be workable (because these are tests) but would be awkward, and some version of `import()` would still be needed.

TimothyGu commented on lib/internal/loader/Loader.js:
```diff
@@ -26,47 +28,77 @@ function getBase() {
+      parentURLOrString = new URL(parentURLOrString);
+    } else if (parentURLOrString instanceof URL === false) {
+      throw new errors.TypeError('ERR_INVALID_ARG_TYPE',
+                                 'parentURLOrString', 'URL');
```
This doesn't really address Domenic's concern of a mutable `URL` object being passed around. Instead you can do `if (typeof parentURLOrString !== 'string' && !(parentURLOrString instanceof URL)) throw ...; const url = new URL(parentURLOrString);`.

domenic replied:
Or just do ToString on any incoming argument like every ES and web API do.

jdalton replied:
I think the Node code precedent here is to use template strings to coerce:
```js
`${url}`
```

guybedford replied:
This resolve isn't exposed publicly, it's the `this.resolver` call just below that represents the hookable API. Will add a guard to that.

jdalton commented:
> having to run it as node --with-loader=./path/to/loader.js ./path/to/test.mjs would basically make that kind of ad hoc command line use too awkward to use.)

`--with-loader=package-name` works too which is similar to how I'm seeing folks use `@std/esm` (using `node -r @std/esm path/to/test.js`)

jdalton commented:
Would that smooth things over?

bcoe commented:
@jdalton @iarna 

> I'd assume --loader package-name works too, which is similar to how I'm seeing folks use @std/esm (using node -r @std/esm ./path/to/test.js). Would that smooth things over?

This becomes a bit awkward for writing testing tools like `nyc`. Users will often spawn a subprocess in a `cwd` that potentially has its own `node_modules` folder. So, unless you were going to require that someone install this category of tool globally (which feels gross), you end up doing a lot of bookkeeping to ensure that spawned subprocesses have the appropriate flags.

This is a similar concern I have with trying to facilitate this functionality using V8's built-in coverage. Which is even more of a Rube Goldberg machine, as you need to spawn a process with `--inspect`, track which port has been used, etc.

jdalton commented:
> Users will often spawn a subprocess in a cwd that potentially has its own node_modules folder. So, unless you were going to require that someone install this category of tool globally (which feels gross)

I don't have a ton of experience with all that. I know that using tools like AVA, which forks processes, I was able to enable `@std/esm` by tying into their `require` option (which is serialized as a JSON string `process.argv` arg). Which is to say if AVA can do it couldn't `nyc` and others too?

bcoe commented:
@jdalton 

> I don't have a ton of experience with all that. I know that using tools like AVA, which forks processes, I was able to enable @std/esm by tying into their require option (which is serialized as a JSON string process.argv arg). Which is to say if AVA can do it couldn't nyc and others too?

nyc already does a lot of this, passing around state so that subprocesses can when spawned have their `require.extensions` patched (so that including a source-file results in it being instrumented by Istanbul).

My concern is that this flow is already very complex and bug prone. As it stands right now the new module system makes writing a development tool like this even more difficult.

bmeck commented:
@bcoe 

> This becomes a bit awkward for writing testing tools like nyc. Users will often spawn a subprocess in a cwd that potentially has its own node_modules folder. So, unless you were going to require that someone install this category of tool globally (which feels gross), you end up doing a lot of bookkeeping to ensure that spawned subprocesses have the appropriate flags.

You can use the `NODE_OPTIONS` environment variable to intercept the subprocess and attach a `--loader` argument. Is that sufficient?

> This is a similar concern I have with trying to facilitate test-coverage functionality using V8's built-in coverage. Which is even more of a Rube Goldberg machine, as you need to spawn a process with --inspect, track which port has been used, etc.

This seems like we can improve that workflow rather than making a more complex hook system.

> My concern is that this flow is already very complex and bug prone. As it stands right now the new module system makes writing a development tool like this even more difficult.

This is exactly why I want to work on a minimal API and iterate rather than trying to recreate the old workflows. If we focus on specific aspects instead of trying to solve all the use cases at once I think we will not suffer the previous problems with standardizing loaders.

bcoe commented:
@bmeck

That's a really thoughtful point of view; mainly I just want to make sure we have examples of how we'd facilitate use-cases like dependency injection, coverage, and transpilation, before ES modules aren't behind a flag... I think if we can nail these three categories of common developer tools, we'd know the API changes are looking really healthy.

bmeck commented:
@bcoe Dependency injection won't work like in other languages where it can change over time due to idempotency of ESM for any given specifier. The spec does not allow things like:

```js
await import('thingFactory') != await import('thingFactory');
```

You can pass in data via query params as needed and that will give you a clean local module map that can change what `thingFactory` is. Can't re-use the same module though.

I'm not going to block unflagging on transpilation as they can be done ahead of time. However, various workflows like needing a way to get coverage is certainly blocking. If that is done ahead of time and works I would not see it as a blocker.

Fishrock123 commented on lib/module.js:
```diff
@@ -490,6 +490,10 @@ function tryModuleLoad(module, filename) {
   }
 }
 
+Module.import = function(request, parent) {
```
Since this is only temporary until `import()` lands, even setting this property should be guarded by the `--experimental-modules` flag.

guybedford commented:
* I've included some more fixes for consistency with respect to no longer passing URLs through APIs.
* I've added "addon" and "json" as `format` values that can be returned by the resolver, making the loading of JSON and native addons independent of the CommonJS module system in the process.

jdalton commented on lib/internal/loader/Loader.js:
```diff
@@ -63,31 +59,27 @@ class Loader {
+    }
+
+    if (!url.startsWith('file:') && !url.startsWith('node:')) {
+      throw new errors.Error('ERR_INVALID_PROTOCOL', url, 'file:');
```
Should there be a separate error for "node:" protocol or is that something internal only that users won't be able to use?

bmeck replied:
i'm not sure / want userland experimentation here. `node:` seems fine, but I don't like occupying a protocol so kinda want it to be semi-discouraged. However, people still need to be able to make their resolver point to the built in modules. I am 50-50 on if removing the `node:` prefix is preferable to keeping it.

I don't want people to use `import 'node:util'` in their ESM right now though. So it is inconsistent which makes me :( , but don't have a better approach.

jdalton replied:
I don't see the `node:` protocol being used for any kind of detection. It's parsed out and ignored as far as I can tell. Why not just use `file:` as with the others now that there is a "addon" hook.

jdalton replied:
Ahh this is for builtin not addon modules. The `node:` prefix made me think it was for `.node` stuff.

jdalton replied:
> However, people still need to be able to make their resolver point to the built in modules

What's the current way Node handles builtin module references as urls?

bmeck replied:
builtin specifiers resolves to an internal `node:$name`, not exposed and does not allow query strings, etc. to builtin modules currently

jdalton replied:
You mean [here](https://github.com/nodejs/node/blob/6c38ab34a301f76a2d5f306ddc242d2bebbd8f53/lib/internal/loader/resolveRequestUrl.js#L58-L78)?

If that's the only place then `node:` can be replace with `file:` since it's not being used *(it's ignored when parsing to just get the pathname)*

guybedford replied:
@jdalton well spotted in that this path is no longer hit as it only applies to `builtin` case. Have removed the node check here. The `node:` scheme is not exposed to users of the resolver as the builtin scenario takes `{ format: 'builtin', url: 'fs' }` with `node:` scheme only used internally for these cases.

jdalton commented on lib/internal/loader/ModuleRequest.js:
```diff
@@ -45,13 +46,41 @@ loaders.set('cjs', async (url) => {
+    debug(`Loading JSONModule ${url}`);
+    const urlObj = new URL(url);
+    const pathname = internalURLModule.getPathFromURL(urlObj);
+    const content = fs.readFileSync(pathname, 'utf8');
```
Style nit: in this case you assign to `urlObj` but in the `.set("cjs")` case you invoke as an argument to `getPathFromURL`.

guybedford replied:
Thanks for review. At one point this was to reuse it, but that changed, certainly can make this consistent.

guybedford commented:
The latest commit here supports the `--loader` definition itself to be a `.mjs` file as a module, which seemed odd from a consistency point of view. The issue with this is it requires making the whole NodeJS bootstrap async, which I've managed to do there, but this will likely need more eyes from those with better knowledge of this area. It would be a nice to have if we can get it in though.

Fishrock123 commented:
Would it be helpful to spec this out a bit on [the eps repo](https://github.com/nodejs/node-eps)?

Might help with some posterity over design choices.

guybedford commented:
@Fishrock123 I was under the impression that repo was deprecated now though?

Fishrock123 commented on lib/internal/bootstrap_node.js:
```diff
@@ -5,11 +5,13 @@
 'use strict';
 
-(function(process) {
+(async function(process) {
```
Does this need a global catch handler now?

guybedford replied:
This gets picked up in node.cc where it is actually executed (as it always has).

The catch handler is then included in the node.cc patch.

Fishrock123 commented:
> @fishrock123 I was under the impression that repo was deprecated now though?

I was unaware of that. I have posted in the related threads that I do not agree.

guybedford commented:
There are still some issues with the async bootstrap approach here unfortunately, if anyone can assist with some advice here I would really appreciate it, otherwise we may just have to accept a CommonJS-based hook process itself.

jdalton commented:
> There are still some issues with the async bootstrap approach here unfortunately,

What are the issues?

guybedford commented:
I've reverted back to CJS-based hooks for now until more progress can be made on getting the async bootstrap to work properly.

@jdalton the issue I've got is I don't know what combination of lifecycle methods to run to allow the async bootstrap to fully unwind itself. My naive attempt here has been moved to a separate branch, where I'm just attempting to run any micro tasks - https://github.com/guybedford/node/blob/async-bootstrap/src/node.cc#L3800. The issue is that the bootstrap promise remains in a pending state and causes a failure at https://github.com/guybedford/node/blob/async-bootstrap/src/node.cc#L3809. I've tried various combinations of lifecycle methods there to try and push this out, but couldn't quite find the right approach.

guybedford commented:
I finally managed to make some progress on the async bootstrap, which is now working for the various lifecycle processes, so we're back on for ES-module-based hooks.

Documentation has been fleshed out, and linting is passing. This is now ready for final review. All feedback welcome.

BridgeAR commented:
This needs a rebase and it should be blocked until we properly test these things on the CI as well.

evanlucas commented:
Thanks for all the effort on this!

Have you by chance checked the performance impact of these changes?

guybedford commented:
@BridgeAR with https://github.com/nodejs/node/pull/15276 landed I believe the CI should be enabled now removing the block? I've rebased, and I can squash whenever ready.

@evanlucas thanks for taking a look, in terms of module loading performance, we haven't reached a point of optimizing the module loading code for performance yet, which I think will be a separate process as things stabilize so I'd be weary of reading too much into any numbers here. This change doesn't affect the algorithm at any deep level though apart from simply how users might intercept the resolver. For the NodeJS bootstrap change to startup time, I've done some basic timing checks on Node startup and get the following over 10 runs running `time ./node --experimental-modules x.mjs` seeing negligible difference:

Current master (ms): 106 105 108 109 105 105 106 108 106 106 (106.4 avg)
With this change (ms): 107 106 106 106 106 105 108 105 106 107 (106.2 avg)

If there is anything else you think might be worth checking let me know.

Fishrock123 commented:
assigning in an attempt to keep track of it, will try to remember on monday

demurgos commented:
> I'm not going to block unflagging on transpilation as they can be done ahead of time. However, various workflows like needing a way to get coverage is certainly blocking. If that is done ahead of time and works I would not see it as a blocker.

Hi,
I'd like to expose my use case for custom loaders. I use Typescript for my Node projects. I compile my application ahead of time and then execute the JS output. The problem is that I also use Typescript for my build scripts. I use both Webpack and Gulp: they do support Typescript entry points. It means that even if I don't need custom loaders to run my application, I use them to build it. Loosing this feature by requiring an "ahead of time" compilation of my `gulpfile.ts` would be a regression. I'd have to either add a pre-build step to compile my gulpfile, drop type checks and revert to using plain JS or keep both the TS source and JS output in source control.

Transpilation on the fly might not block unflagging, but I hope that support for it is still planned.

Support for ES modules by TS-Node is discussed in [this issue](https://github.com/TypeStrong/ts-node/issues/436).

bmeck commented:
@demurgos is there any problem with my mention that you can do it "on the fly" by creating a new URL with the content body you need (by making a file or using things like URL.createObjectURL [separate PR will come for that])?

Fishrock123 commented on src/node.cc:
```diff
@@ -3750,16 +3755,35 @@ void LoadEnvironment(Environment* env) {
-  if (ret.IsEmpty())
-    env->async_hooks()->clear_async_id_stack();
+
+  if (f->IsAsyncFunction()) {
```
is this branch missing `env->async_hooks()->clear_async_id_stack();`?

guybedford replied:
Good point, I missed this detail on rebase, will add this to the error path.

Fishrock123 commented on src/node.cc:
```diff
@@ -3750,16 +3755,35 @@ void LoadEnvironment(Environment* env) {
+        f->Call(Null(env->isolate()), 1, &arg).As<Promise>();
+
+    // bootstrap event loop
+    while (bootstrap_promise->State() == Promise::kPending) {
```
This seems risky?

What if a commonjs module with first-tick eventloop work is loaded in the tree? Might it execute async code before load is finished?

Also, are you sure this ever returns, instead of just looping here?

guybedford replied:
Since bootstrap code is all known code this doesn't seem like a problematic constraint to me, that the bootstrap promise always resolves. If you know a better way to handle this based on the right checks to see if work is still happening through this loop I'm all ears. It's only indefinitely pending promises that might cause these issues though - error cases all work out fine as they always reject the top-level promise (I have tested things like maximum call stack on bootstrap etc).

> What if a commonjs module with first-tick eventloop work is loaded in the tree? Might it execute async code before load is finished?

Do you have any examples of where first-tick-scheduled work is specifically needed through the bootstrap? If there are cases like this, it may make sense to delegate them to an init-based tick or similar perhaps rather.

demurgos commented:
> @demurgos is there any problem with my mention that you can do it "on the fly" by creating a new URL with the content body you need (by making a file or using things like URL.createObjectURL [separate PR will come for that])?

Thanks: I had to re-read your comment but it would work for "fetch transformations" such as transpilation. My biggest issue was that I wasn't sure how transitive dependencies (all of them requiring transpilation) would be handled, but it should work out of the box with this solution (`resolve` simply returning an URL for the compiled output).

bcoe commented:
@bmeck would the `URL.createObjectURL` API essentially allow us to transpile the code at the path that would have been loaded, and return the alternative code in a cached blob?

bmeck commented:
@bcoe https://developer.mozilla.org/en-US/docs/Web/API/URL/createObjectURL , it would let you resolve to a new URL with an alternate source text, yes. It also solves the inconsistency of replacing the URL in the process for code loading vs not replacing it in the process when using `fs`.

Fishrock123 commented:
Honestly, this is like double-flagged experimental and it seems fine enough to me. I really don't have enough opinions about this.

The async-ifying of bootstrap does make me very cautious though - it seems to me like there are any number of things soon waiting to go wrong at that level.

guybedford commented:
Thanks @Fishrock123, I will certainly keep an eye on the async bootstrap code - this isn't just something I'm aiming to commit and walk away from too.

I've rebased the PR here, would be great to get this merged!

bmeck commented:
CI https://ci.nodejs.org/job/node-test-pull-request/10534/

Going to merge this at end of day if all goes well in CI

gibfahn commented:
Had to kill CI after 4 hours and 10 minutes. It looks like something went wrong:

<img width="646" alt="image" src="https://user-images.githubusercontent.com/15943089/31349513-657fcfd2-acd8-11e7-9452-382163d77176.png">


Also there were a lot of failures, e.g. https://ci.nodejs.org/job/node-test-commit-linux/nodes=fedora25-x64/12986/console

bmeck commented:
lets definitely hold off until CI looks green

guybedford commented:
I'm looking into this now.

@Fishrock123 well, had to keep an eye on it sooner than expected :)

guybedford commented:
So it turns out the async bootstrap has been too much of a hassle to try and add - I was hoping the right approach might have come along, but it didn't. If the bootstrap does ever get updated to be async, we could potentially come back to this approach.

Instead I've moved the hooking into a check on module load rather. CI should be good now.

gibfahn commented:
CI 2: https://ci.nodejs.org/job/node-test-commit/12956/

bmeck commented:
CI 3: https://ci.nodejs.org/job/node-test-commit/12988/

bmeck commented:
CI is passing, will merge tomorrow morning if no objections.

targos commented:
why dont-land-on-v8.x?

targos commented on lib/internal/bootstrap_node.js:
```diff
@@ -5,6 +5,8 @@
 // to the performance of the startup process, many dependencies are invoked
 // lazily.
 
+/* eslint-disable no-extra-parens */
```
I suppose this can be removed?

bmeck commented:
@targos it was modifying node bootstrap to be async when I added that. we can remove it now that we removed the async bootstrap

targos commented on src/node.cc:
```diff
@@ -3949,7 +3955,9 @@ static void CheckIfAllowedInEnv(const char* exe, bool is_env,
     "--pending-deprecation",
     "--no-warnings",
     "--napi-modules",
-    "--expose-http2",   // keep as a non-op through v9.x
```
the comment should not be removed

targos commented on lib/internal/loader/ModuleJob.js:
```diff
@@ -2,51 +2,40 @@
-      this.instantiated = this.modulePromise;
-    }
+    // instantiated == deep dependency jobs wrappers instantiated,
+    //module wrapper instantiated
```
can you add a space after the slashes?

bmeck commented:
Landed in d21a11dc23d6104b1d03fa2ddc1c808dcaf89c31

Trott commented:
Unfortunately, it appears this landed without a CI run 😱 and it causes Raspberry Pi devices to fail:

https://ci.nodejs.org/job/node-test-binary-arm/10851/RUN_SUBSET=4,label=pi1-raspbian-wheezy/console

```console
not ok 249 es-module/test-esm-addon
  ---
  duration_ms: 2.65
  severity: fail
  stack: |-
    (node:29865) ExperimentalWarning: The ESM module loader is experimental.
    Error: module ../addons/hello-world/build/Release/binding.node not found
        at module.exports (internal/loader/search.js:14:12)
        at exports.resolve (internal/loader/ModuleRequest.js:91:13)
        at Loader.resolve (internal/loader/Loader.js:50:40)
        at Loader.getModuleJob (internal/loader/Loader.js:78:40)
        at ModuleWrap.module.link (internal/loader/ModuleJob.js:27:25)
        at linked (internal/loader/ModuleJob.js:25:19)
        at <anonymous>
```

Trott commented:
Oh, wait, I'm totally wrong, it had CI, even passing CI. (I searched for `CI:` rather than just `CI` because I'm a chump.) Sorry about that.

OK, so wonder what changed to cause the rash of failures all of a sudden...

bcoe commented:
@bmeck where would be a good place to start (or continue) a discussion about `createObjectURL`; I think this would be the most promising short term solution for keeping nyc/Istanbul working as we start to move some of the Node core modules over to `.mjs`.

Trott commented:
Talking with @refack in IRC. The issue is indeed with this change but CI didn't catch it because of a bug in Makefile that caused the test to be omitted on CI. 

Most relevant bit:

> Yeah. This can't work on the ARM https://github.com/nodejs/node/blob/537402f09590f088bf1cde063f213d45e0d6e4e8/test/es-module/test-esm-addon.mjs#L5
> And is also build-order-sensitive on all other platforms
```

Trott commented:
Ref: https://github.com/nodejs/node/issues/16155#issuecomment-336229318

refack commented:
> Yeah. This can't work on the ARM
> And is also build-order-sensitive on all other platforms

For reference the `arm-fanned` job runs the addons suite on a different machine than any of the JS tests, so there is no compilation of the `hello-world` addon before `node/test/es-module/test-esm-addon.mjs` is run. (and also on the other platforms it's only by chance that it is done before this test is run).
tl;dr is should be moved to `test/addons/hello-world/`

Trott commented:
Proposed fix in https://github.com/nodejs/node/pull/16174
